### PR TITLE
Azure Monitor: Add search feature to resource picker

### LINKED ISSUE #46063: Azure Monitor: Add support for filtering on the Resource Picker
Nice to have: Add support for filtering existing loaded resources on the frontend.

### PR DESCRIPTION
**Pull request by sarahzinger**
**What this PR does / why we need it**:

Adds search functionality for the resource picker for azure monitor.

**Which issue(s) this PR fixes**:
Fixes #46063

**Special notes for your reviewer**:
- I also fixed an unrelated bug here for scrolling resources into view
- I did some additional refactoring to make our resource picker less "smart" and our resourcePickerData class to be more "smart", if this is feeling like too much change at once let me know and I can break it up into a few smaller prs.
- these queries will match for anything where the search matches the uri, which is...maybe a bit unexpected and it should instead just match by name? But when I look at the network tab of azure portal to see how it's done with their resource picker, that seems to be the way to do it, which I imagine is useful when searching by id. It seemed reasonable to start with matching behavior as much as we think makes sense and go from there. 
- I didn't bother trying to reconstruct a tree as it looks like they do not do that in azure portal and it seemed maybe more trouble than its worth but I could see that being a feature in the future if we want?
- I was thinking of putting this in 9.0 but if it has these bug fixes maybe it should be in 8.5.x, I guess this is another argument for me to break this pr up into multiple ones. I guess let me know what you think. I don't think it should be breaking changes so it could go in an 8.5 release, but it's a bit strange to put new features in a patch.

### CODE REVIEW

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/resourcePicker/resourcePickerData.ts
Comment (andresmgot) [inline public/app/plugins/datasource/grafana-azure-monitor-datasource/resourcePicker/resourcePickerData.ts:39]: a more readable interface for this function would be `fetchInitialRows(subscription?: string, resourceGroup?: string)`. That way it's more clear that depending on what you pass you will get a different set of resources.

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/resourcePicker/resourcePickerData.ts
Comment (andresmgot) [inline public/app/plugins/datasource/grafana-azure-monitor-datasource/resourcePicker/resourcePickerData.ts:67]: nit: this method is making it more difficult to know what's happening. Since it's only used once I would avoid it. Also, let's avoid the name ` resourceGroupOrSubscription` so we don't need to add more `Or...` in the future :sweat_smile:

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/components/ResourcePicker/ResourcePicker.tsx
Comment (andresmgot) [inline public/app/plugins/datasource/grafana-azure-monitor-datasource/components/ResourcePicker/ResourcePicker.tsx:112]: I don't think you are supposed to set the loading status to 'NotStarted' since that should be only the default state. IMO, I would just make the LoadingStatus a boolean and either check in the line 48 if the status is undefined or simply change that `useEffect` to `useEffectOnce` since it's meant to be run only once.

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/resourcePicker/resourcePickerData.ts
Comment (andresmgot) [inline public/app/plugins/datasource/grafana-azure-monitor-datasource/resourcePicker/resourcePickerData.ts:77]: curious, why the limit to 200?

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/resourcePicker/resourcePickerData.ts
Comment (andresmgot) [inline public/app/plugins/datasource/grafana-azure-monitor-datasource/resourcePicker/resourcePickerData.ts:71]: I don't fully understand this `if`, can you comment on it?

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/resourcePicker/resourcePickerData.ts
Comment (andresmgot) [inline public/app/plugins/datasource/grafana-azure-monitor-datasource/resourcePicker/resourcePickerData.ts:108]: can you add tests for the new functions here?

**Review by andresmgot — COMMENTED** [review]
Great addition! Love the feature :rocket:.

re: Split the PR

Usually, when you wonder if a PR should be split, it means that it should be split :) The quality of the review is highly reduced (and the time needed to review it highly increases). Basically the reviewer end up trying to run the code without actually taking a deeper look to it. At the end the price is paid with less quality code.

In this case, I agree removing at least the "scroll into view" change. Mostly because it's not clear now what was the issue and how it was fixed (and also to include the fix in 8.5).

Apart from that I found a bug: If there is an element selected and you do a search that does not match with your resource, the selection gets lost, allowing you to select a new resource. That doesn't happen if the search matches the current selection. The current selection should not be lost when using the search bar.

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/resourcePicker/resourcePickerData.ts
Comment (sarahzinger) [reply public/app/plugins/datasource/grafana-azure-monitor-datasource/resourcePicker/resourcePickerData.ts:67]: my reason for moving it in the resourcePickerData was not because I foresee us using it again in some other way (none of the methods in here are used all that frequently), but because I think its helpful to separate out our code by functionality, we have a resourcePickerData file, that I think is responsible for fetching and parsing data, and a resourcePicker react component that is responsible for rendering that parsed data and handling user interactions.

I think having everything in one pr makes it difficult to see :D so I took your advice and tried to move this sort of thing in a separate pr: https://github.com/grafana/grafana/pull/48312 I also changed the argument names rather than just copy paste what we had before, let me know what you think if you're interested!

**Comment by sarahzinger** [comment]
cool thanks andres :) I'm moving this to a draft while i edit. In the meantime here are 2 prs that I took from this one to make things easier to look at:
https://github.com/grafana/grafana/pull/48311
https://github.com/grafana/grafana/pull/48312

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/components/ResourcePicker/ResourcePicker.tsx
Comment (sarahzinger) [reply public/app/plugins/datasource/grafana-azure-monitor-datasource/components/ResourcePicker/ResourcePicker.tsx]: So when you type in a search, we erase the old contents, and if you enter backspace so that there is no longer a search phrase, we have to reset the resource picker to have some kind of results. There are a couple of ways that I could think of trying that:
1. cache the old results in a state variable and reset them back to that (this seemed bad to me because we risk stale data)
2. we could just return whatever search gives for an empty search phrase (a flattened list of everything) that seemed confusing to me and would probably cause timeouts.
3. we could retrigger the same call we do on mount which is to get all subscriptions, and only get resource groups or resources if there was a previous selection, and then only for those opened folders.

I figure 3 seemed like a reasonable response so need some way to restart the useEffect not just on mount but when the search field goes from something to nothing, so that's what's happening here. That all said "NotStarted" is not an accurate description of this either, I renamed some stuff if you want to take a look and let me know if it seems clear and like a good approach.

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/resourcePicker/resourcePickerData.ts
Comment (sarahzinger) [reply public/app/plugins/datasource/grafana-azure-monitor-datasource/resourcePicker/resourcePickerData.ts]: I believe this is what Azure Portal does, it seemed like a reasonable number to start with?

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/resourcePicker/resourcePickerData.ts
Comment (sarahzinger) [reply public/app/plugins/datasource/grafana-azure-monitor-datasource/resourcePicker/resourcePickerData.ts]: yeahhhhh, this is...unnecessarily confusing. Essentially I'm trying to say if you can select only resources, then only search for resources (aka metrics queries), if you can select for resources and resource groups, then search for resources and resources groups (aka logs queries).

I revised this to hopefully make it more clear let me know what you think!

**Review by sunker — COMMENTED** [review]
This is really exciting! 

I know this PR is a draft but wanted to give some high level comments. If addressed, they could be added step by step and not necessarily in now. 

>Fixes https://github.com/grafana/grafana/issues/46063

Seems like this solves a different problem - search and not filtering? 

>these queries will match for anything where the search matches the uri, which is...maybe a bit unexpected and it should instead just match by name?

Yeah that was my initial thought - that it's a bit weird to get matching resources where the search phrase is never to be seen in the resource picker list. I think I'd prefer to let the default behavior be to search for the resource name, and then in the future we could make it a search option to change that to search by other fields such as id, type etc. 

>I didn't bother trying to reconstruct a tree as it looks like they do not do that in azure portal and it seemed maybe more trouble than its worth but I could see that being a feature in the future if we want?

I agree, this is better. 


For us, the resource graph api is blending fast but we also don't have that many resources. I wonder how the search feature would work in case ARG is much slower than it is for us? Also not sure if ARG it's a part of the free-tier or if it cost money? If slow or pricey, it might not be ideal to issue a new query meanwhile typing (although it's currently with a debounce). I think we should consider having a search button. If you want to, you could demo this in a UX feedback session. They might have good feedback to give and also let us now if this search feature is aligned with other search implementations in grafana.

I think limiting the result to something like 200 results is good since we've had issues in the past where the browser crashes due to out of memory because we're loading too much data from ARG. However, I wonder if it would be better to set that limit in the [query options](https://docs.microsoft.com/en-us/azure/governance/resource-graph/concepts/work-with-data#paging-results) instead. That way the returned value for `totalRecords` would still be accurate, and we could display something like this is a label: _Showing first 200 resources out of 800 matching results_.

**Comment by andresmgot** [comment]
> > Fixes #46063
> 
> Seems like this solves a different problem - search and not filtering?

Yes, the reasoning behind this is that at first we thought of just filtering whatever was already loaded. That has its disadvantages so we ended up doing an active search (but didn't update the issue). 

> > these queries will match for anything where the search matches the uri, which is...maybe a bit unexpected and it should instead just match by name?
> 
> Yeah that was my initial thought - that it's a bit weird to get matching resources where the search phrase is never to be seen in the resource picker list. I think I'd prefer to let the default behavior be to search for the resource name, and then in the future we could make it a search option to change that to search by other fields such as id, type etc.
> 

I agree it may be a bit weird to get everything but...
 - Azure portal is doing the same thing
 - It's a legit use case to say "I want to search for resources in my foobar resource group so I will search that"

The downside is that if you look for example for "subscription" you get all the resources (since all the uris match that) but again that's how the Azure portal behaves.

> For us, the resource graph api is blending fast but we also don't have that many resources. I wonder how the search feature would work in case ARG is much slower than it is for us? Also not sure if ARG it's a part of the free-tier or if it cost money? If slow or pricey, it might not be ideal to issue a new query meanwhile typing (although it's currently with a debounce). I think we should consider having a search button. If you want to, you could demo this in a UX feedback session.

Here I'd argue not to put the bandage before the wound. I cannot find any documentation about ARG costing money (it makes sense that you pay for the resources deployed, not for listing them). Having a button would hurt (just a bit) the UX and may only be helpful for some cases (but I don't think so because of the 200 limit) so I think it's fine as it is.

> However, I wonder if it would be better to set that limit in the [query options](https://docs.microsoft.com/en-us/azure/governance/resource-graph/concepts/work-with-data#paging-results) instead. That way the returned value for totalRecords would still be accurate, and we could display something like this is a label: Showing first 200 resources out of 800 matching results.

That'd be a nice to have (but I wouldn't worry too much about it). Usually if you get a vague response, you add more details to the search term. It'd be nice to know that the response that you get is partial though in case someone inspects the 200 items and don't find what they were looking for.

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/components/ResourcePicker/ResourcePicker.tsx
Comment (andresmgot) [reply public/app/plugins/datasource/grafana-azure-monitor-datasource/components/ResourcePicker/ResourcePicker.tsx]: gotcha, yes I agree with 3 there but modifying the loading status to trigger a fetch seems unnecessarily complex. That use effect even already defines the function loadInitialData. You just need to call it there and here so your intention is explicit.

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/resourcePicker/resourcePickerData.ts
Comment (andresmgot) [inline public/app/plugins/datasource/grafana-azure-monitor-datasource/resourcePicker/resourcePickerData.ts:107]: nit: I find this a bit more readable but up to you

```suggestion
      let id = parsedUri.subscriptionID;
      let type = ResourceRowType.Subscription;
      if (parsedUri.resource) {
        id = parsedUri.resource;
        type = ResourceRowType.Resource;
      } else if (parsedUri.resourceGroup) {
        id = parsedUri.resourceGroup;
        type = ResourceRowType.ResourceGroup;
      }
      return {
        name: item.name,
        id,
        uri: item.id,
        resourceGroupName: item.resourceGroup,
        type,
        typeLabel: resourceTypeDisplayNames[item.type] || item.type,
        location: locationDisplayNames[item.location] || item.location,
      };
```

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/resourcePicker/resourcePickerData.ts
Comment (andresmgot) [inline public/app/plugins/datasource/grafana-azure-monitor-datasource/resourcePicker/resourcePickerData.ts:73]: I get why the query is different for metrics and logs but maybe it shouldn't? Checking the Azure portal again, it seems that the behavior there is to include resource groups (and subs) there. The case is different because there it allows you to further select a resource type and query multiple resources (which we don't support for metrics):

Screenshot from 2022-04-28 12-48-36

I would re-use the logs query because:
 - It is get us closer to the portal approach, if we add support to query multiple resources for metrics
 - If someone was looking for a resource-group (even if they cannot use it), it would be confusing to skip that value.

So for the current state, I would list the resource but make it non selectable. Does that make sense?

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/components/ResourcePicker/Search.tsx
Comment (andresmgot) [inline public/app/plugins/datasource/grafana-azure-monitor-datasource/components/ResourcePicker/Search.tsx:9]: nit: I would increase the debounce time a bit (like 500-600ms)

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/resourcePicker/resourcePickerData.ts
Comment (andresmgot) [reply public/app/plugins/datasource/grafana-azure-monitor-datasource/resourcePicker/resourcePickerData.ts]: commented below

**Comment by sunker** [comment]
> The downside is that if you look for example for "subscription" you get all the resources (since all the uris match that) but again that's how the Azure portal behaves.

This is totally fine for a first version of this feature. Basing the search on other fields could be a feature for the future because it would add a lot of value. 

>Having a button would hurt (just a bit) the UX and may only be helpful for some cases (but I don't think so because of the 200 limit) so I think it's fine as it is.

Respectfully disagree here. :) I think that it's expected that a "search feature" has a search button. "Filters" normally don't. The resource picker in the Portal is not doing this well IMO. Making the API request on button click instead of text onchange would be the easy way of solving this. Anyway, would be great to run this through UX in a feedback session. 

>That'd be a nice to have (but I wouldn't worry too much about it). Usually if you get a vague response, you add more details to the search term. It'd be nice to know that the response that you get is partial though in case someone inspects the 200 items and don't find what they were looking for.

This could also come in future releases, but I think a label such as _Showing first 200 resources out of 800 matching results_ would be low effort but still add a lot of value to the user.

**Comment by sarahzinger** [comment]
Hi @andresmgot and @sunker, thank you for the reviews and helpful feedback! I've made some changes and also thought about the suggestions and some thoughts in response in one comment I hope that's ok!

**Search Vs Filtering**
- sorry the issue for that is out of date! I had gone back and forth a few times on this. We landed on searching because:
	- if you filter for something that exists but is not loaded, its an odd experience. 
	- search seems potentially more useful and saves you from having to make a million arg calls if you don't remember which resource group your resource is hiding in. 
	- azure portal uses search and it seems not really any harder to implement than filtering
**Searching for uri vs searching for name.** 
	- this one feels tricky. I could really see someone wanting to search by an id and a search by name won't necessarily help, but searching by resource uri would. I think for maybe this iteration it makes sense to match azure portal's functionality for this and then iterate if we can with more advanced filtering if there seems to be need. I can make an discussion for this after I merge if that sounds good?
**Can we demo this at a ux session?**
	- yes! I'll sign up, if it's cool I'd like to merge it before then but maybe if there's feedback we can make issues and fix them in future iterations, does that sound ok?
**Debounce vs search button.** 
	- I am partial to this debounce :) it's how azure portal works and i think it's not _terribly_ worse in performance or different than clicking open all these folders. I also think more and more of the web is moving towards this as a default functionality. That said I don't want our users to be charged a bunch or something by accident! I will bring it up at the ux session and it seems like a quick fix to add the button if we need. In the meantime I slowed down the debounce and it seems reasonable, if that's ok! 
**200 records- is it enough? Can we show the total count like "_Showing first 200 resources out of 800 matching results_."**
	- So interestingly totalRecords and count seem to be limited to the limit you pass in. So if you say the limit is 10 and you request something that has more than 10 results, in the response back both totalRecords and count will read as 10. I don't really understand why :( So unfortunately without that data I don't think it's possible to show that kind of message (without maybe making a separate count call?, but if I'm missing something let me know as I agree this would be great to show users.
**Can we use query options**
	- sounds like a great feature request for the future! I can be sure to make an issue for this after I merge
**Making the same Logs search for metrics queries**
	- Back in our planning days, I think we said that we weren't quite ready to handle multi select and wanted to hold off on that which is why I didn't. 
	- And If we load non selectable items in a flattened structure like this, I think it'll be pretty frustrating to users as there's not really any point to them except to say "yes you have these resources and no we're not letting you do anything with them". Seems easy to implement though, so if you feel strongly about it, I'm happy to do it. I just wonder if it would be more frustrating then helpful? We could do a search and maintain a tree like structure but I think that would require either making multiple search queries or some very fancy tree sorting like code on our end. Either way I kinda feel like that's beyond this pr? What do you think?

**Comment by andresmgot** [comment]
> **Searching for uri vs searching for name.** - this one feels tricky. I could really see someone wanting to search by an id and a search by name won't necessarily help, but searching by resource uri would. I think for maybe this iteration it makes sense to match azure portal's functionality for this and then iterate if we can with more advanced filtering if there seems to be need. I can make an discussion for this after I merge if that sounds good?

I agree with this.

> **Debounce vs search button.** - I am partial to this debounce :) it's how azure portal works and i think it's not _terribly_ worse in performance or different than clicking open all these folders. I also think more and more of the web is moving towards this as a default functionality. That said I don't want our users to be charged a bunch or something by accident! I will bring it up at the ux session and it seems like a quick fix to add the button if we need. In the meantime I slowed down the debounce and it seems reasonable, if that's ok!

I also agree with this. I guess that the discussion about using a button rather than auto-search is based in two arguments:

 - Search may cost money. It's not the case. See this [docs](https://docs.microsoft.com/en-us/azure/governance/resource-graph/overview#throttling): "As a free service, queries to Resource Graph are throttled".
 - Search may be slow. It looks like it's not the case either. We are limiting the response to 200 results so regardless of the number of resources in the account, the performance improvements should be made in the API side so it should be fast enough.
 
In any case, it would be good to check this with UX so we are not missing anything.

> **200 records- is it enough? Can we show the total count like "_Showing first 200 resources out of 800 matching results_."** - So interestingly totalRecords and count seem to be limited to the limit you pass in. So if you say the limit is 10 and you request something that has more than 10 results, in the response back both totalRecords and count will read as 10. I don't really understand why :( So unfortunately without that data I don't think it's possible to show that kind of message (without maybe making a separate count call?, but if I'm missing something let me know as I agree this would be great to show users.
> 
> **Can we use query options** - sounds like a great feature request for the future! I can be sure to make an issue for this after I merge

This is fairly simple, it could be added to this PR. You only need to remove the `| limit 200` from the Kusto query and use the [`$top` query option](https://docs.microsoft.com/en-us/rest/api/azureresourcegraph/resourcegraph(2021-03-01)/resources/resources#queryrequestoptions). I checked this with a `curl` request and seems to work as expected (at the end of the query body you can see `"options":{"resultFormat":"objectArray","$top":"10"}}`:

```bash
▶ curl 'http://localhost:3000/api/datasources/42/resources/resourcegraph/providers/Microsoft.ResourceGraph/resources?api-version=2021-03-01' \
  -H 'Accept-Language: en-US,en;q=0.9' \
  -H 'Connection: keep-alive' \
  -H 'Cookie: grafana_session=' \
  -H 'Origin: http://localhost:3000' \
  -H 'Referer: http://localhost:3000/d/GpllKjeMk/azure-monitor?orgId=1&editPanel=5' \
  -H 'Sec-Fetch-Dest: empty' \
  -H 'Sec-Fetch-Mode: cors' \
  -H 'Sec-Fetch-Site: same-origin' \
  -H 'accept: application/json, text/plain, */*' \
  -H 'content-type: application/json' \
  -H 'x-grafana-org-id: 1' \
  --data-raw '{"query":"\n        resources\n        | union resourcecontainers \n        | where id contains \"subscription\"\n        | where type in (\"microsoft.analysisservices/servers\",\"microsoft.synapse/workspaces/bigdatapools\",\"microsoft.apimanagement/service\",\"microsoft.appconfiguration/configurationstores\",\"microsoft.web/sites/slots\",\"microsoft.web/hostingenvironments\",\"microsoft.web/serverfarms\",\"microsoft.web/sites\",\"microsoft.network/applicationgateways\",\"microsoft.insights/components\",\"microsoft.automation/automationaccounts\",\"microsoft.insights/autoscalesettings\",\"microsoft.aadiam/azureadmetrics\",\"microsoft.cache/redis\",\"microsoft.documentdb/databaseaccounts\",\"microsoft.kusto/clusters\",\"microsoft.dbformariadb/servers\",\"microsoft.dbformysql/servers\",\"microsoft.dbforpostgresql/flexibleservers\",\"microsoft.dbforpostgresql/servergroupsv2\",\"microsoft.dbforpostgresql/servers\",\"microsoft.dbforpostgresql/serversv2\",\"microsoft.resources/subscriptions\",\"microsoft.appplatform/spring\",\"microsoft.databoxedge/databoxedgedevices\",\"microsoft.azurestackresourcemonitor/storageaccountmonitor\",\"microsoft.synapse/workspaces\",\"microsoft.network/bastionhosts\",\"microsoft.batch/batchaccounts\",\"microsoft.botservice/botservices\",\"microsoft.netapp/netappaccounts/capacitypools\",\"microsoft.classiccompute/domainnames\",\"microsoft.vmwarecloudsimple/virtualmachines\",\"microsoft.cognitiveservices/accounts\",\"microsoft.network/networkwatchers/connectionmonitors\",\"microsoft.network/connections\",\"microsoft.containerinstance/containergroups\",\"microsoft.containerregistry/registries\",\"microsoft.insights/qos\",\"microsoft.customerinsights/hubs\",\"microsoft.datafactory/datafactories\",\"microsoft.datafactory/factories\",\"microsoft.datalakeanalytics/accounts\",\"microsoft.datalakestore/accounts\",\"microsoft.datashare/accounts\",\"microsoft.synapse/workspaces/sqlpools\",\"microsoft.devices/provisioningservices\",\"microsoft.compute/disks\",\"microsoft.network/dnszones\",\"microsoft.eventgrid/domains\",\"microsoft.eventgrid/systemtopics\",\"microsoft.eventgrid/topics\",\"microsoft.eventhub/clusters\",\"microsoft.eventhub/namespaces\",\"microsoft.network/expressroutecircuits\",\"microsoft.network/expressrouteports\",\"microsoft.network/expressroutegateways\",\"microsoft.fabric.admin/fabriclocations\",\"microsoft.network/azurefirewalls\",\"microsoft.network/frontdoors\",\"microsoft.hdinsight/clusters\",\"microsoft.storagecache/caches\",\"microsoft.logic/integrationserviceenvironments\",\"microsoft.iotcentral/iotapps\",\"microsoft.devices/iothubs\",\"microsoft.keyvault/vaults\",\"microsoft.kubernetes/connectedclusters\",\"microsoft.containerservice/managedclusters\",\"microsoft.media/mediaservices/liveevents\",\"microsoft.network/loadbalancers\",\"microsoft.operationalinsights/workspaces\",\"microsoft.logic/workflows\",\"microsoft.machinelearningservices/workspaces\",\"microsoft.media/mediaservices\",\"microsoft.network/natgateways\",\"microsoft.network/networkinterfaces\",\"microsoft.network/networkvirtualappliances\",\"microsoft.network/networkwatchers\",\"microsoft.notificationhubs/namespaces/notificationhubs\",\"microsoft.network/p2svpngateways\",\"microsoft.peering/peeringservices\",\"microsoft.powerbidedicated/capacities\",\"microsoft.network/privateendpoints\",\"microsoft.network/privatelinkservices\",\"microsoft.network/publicipaddresses\",\"microsoft.cache/redisenterprise\",\"microsoft.relay/namespaces\",\"microsoft.synapse/workspaces/scopepools\",\"microsoft.search/searchservices\",\"microsoft.servicebus/namespaces\",\"microsoft.signalrservice/signalr\",\"microsoft.operationsmanagement/solutions\",\"microsoft.sql/servers/databases\",\"microsoft.sql/servers/elasticpools\",\"microsoft.sql/managedinstances\",\"microsoft.storage/storageaccounts\",\"microsoft.classicstorage/storageaccounts\",\"microsoft.storagesync/storagesyncservices\",\"microsoft.streamanalytics/streamingjobs\",\"microsoft.media/mediaservices/streamingendpoints\",\"microsoft.timeseriesinsights/environments\",\"microsoft.network/trafficmanagerprofiles\",\"microsoft.compute/virtualmachinescalesets\",\"microsoft.compute/virtualmachines\",\"microsoft.classiccompute/virtualmachines\",\"microsoft.network/virtualnetworkgateways\",\"microsoft.netapp/netappaccounts/capacitypools/volumes\",\"microsoft.network/vpngateways\",\"microsoft.cdn/cdnwebapplicationfirewallpolicies\",\"microsoft.web/hostingenvironments/workerpools\",\"microsoft.resources/subscriptions/resourcegroups\")\n        | order by tolower(name) asc\n        ","options":{"resultFormat":"objectArray","$top":"10"}}' \
  --compressed | jq .               
{
  "totalRecords": 581,
  "count": 10,
  "data": [
```

What would be interesting is to compare the performance of both queries. I am not sure if the performance of limiting the result to 200 in the Kusto query performs better than using the `$top` parameter. I would assume so since the later scan all the resources but in our dev environment (~600 resouces) the difference is not significant.

So, to be safe, I think a good compromise would be to check the size of the response using `| limit 200`. If the size is 200, then we can add a note at the end of the list saying: _Showing only the first 200 resources matching the search term_. WDYT?

**Comment by andresmgot** [comment]
> **Making the same Logs search for metrics queries**
> - Back in our planning days, I think we said that we weren't quite ready to handle multi select and wanted to hold off on that which is why I didn't.

Right, I am not saying that we should handle multi-select here, just that using the same query would be a good step in that direction.

> - And If we load non selectable items in a flattened structure like this, I think it'll be pretty frustrating to users as there's not really any point to them except to say "yes you have these resources and no we're not letting you do anything with them". Seems easy to implement though, so if you feel strongly about it, I'm happy to do it. I just wonder if it would be more frustrating then helpful? 

Not feeling strongly about this one. I believe that there is some UX principle to avoid hiding stuff from users but for sure there is also for not unnecessary clutter the UI :) Maybe a nice thing to check with UX people.

> We could do a search and maintain a tree like structure but I think that would require either making multiple search queries or some very fancy tree sorting like code on our end. Either way I kinda feel like that's beyond this pr? What do you think?

yeah, that's not something I would add to this PR.

**Review by andresmgot — APPROVED** [review]
My +1 to merge this as it is! The only thing I would add before is the note in case a search hits the 200 limit.

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/components/ResourcePicker/ResourcePicker.tsx
Comment (kevinwcyu) [inline public/app/plugins/datasource/grafana-azure-monitor-datasource/components/ResourcePicker/ResourcePicker.tsx:170]: When I use VoiceOver, this doesn't get announced. I think we can add an `aria-live="polite"` here to announce when no results are found, although I'm not actually sure if that's the correct way to do it.

**Review by kevinwcyu — APPROVED** [review]
🔥 so much nicer not having to click through so many levels to select a resource.

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/components/ResourcePicker/ResourcePicker.tsx
Comment (sarahzinger) [reply public/app/plugins/datasource/grafana-azure-monitor-datasource/components/ResourcePicker/ResourcePicker.tsx]: Yeah, I'm not entirely sure either but in manually testing it out it seems reasonable to me!

**Pull request merged**
